crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -1,12 +1,98 @@
+    group, if_group_breaks, soft_block_indent, soft_line_break_or_space, text,
+};
+use ruff_formatter::{format_args, write, Buffer, Format, FormatResult};
+use ruff_python_ast::prelude::{Expr, Ranged};
+use ruff_text_size::{TextLen, TextRange};
 use rustpython_parser::ast::ExprTuple;
 
 #[derive(Default)]
 pub struct FormatExprTuple;
 
 impl FormatNodeRule<ExprTuple> for FormatExprTuple {
     fn fmt_fields(&self, item: &ExprTuple, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(f, [verbatim_text(item.range)])
+        let ExprTuple {
+            range,
+            elts,
+            ctx: _,
+        } = item;
+
+        let saved_level = f.context().node_level();
+        // Tell the children they need parentheses
+        f.context_mut().set_node_level(NodeLevel::Parenthesized);
+
+        let expr_sequence = ExprSequence::new(elts);
+        let inner = soft_block_indent(&expr_sequence);
+        // If the tuple has parentheses, keep them. Also only top level tuples are allowed to elide
+        // them (TODO(konstin): this may be redundant)
+        if is_parenthesized(*range, elts, f) || saved_level != NodeLevel::TopLevel {
+            write!(f, [group(&format_args![&text("("), inner, &text(")"),])])?;
+        } else {
+            write!(
+                f,
+                [group(&format_args![
+                    if_group_breaks(&text("(")),
+                    inner,
+                    if_group_breaks(&text(")")),
+                ])]
+            )?;
+        }
+
+        f.context_mut().set_node_level(saved_level);
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct ExprSequence<'a> {
+    elts: &'a [Expr],
+}
+
+impl<'a> ExprSequence<'a> {
+    const fn new(elts: &'a [Expr]) -> Self {
+        Self { elts }
     }
 }
+
+impl Format<PyFormatContext<'_>> for ExprSequence<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        // If we only have a single entry, we need to add a trailing comma or the tuple will become
+        // an expression
+        if let [single_entry] = self.elts {
+            write!(f, [single_entry.format(), text(",")])?;
+            return Ok(());
+        }
+        let mut iter = self.elts.iter();
+        // Write one comma less than we have elements in the list (i.e. trailing or preceding comma)
+        if let Some(first) = iter.next() {
+            first.format().fmt(f)?;
+        }
+        for entry in iter {
+            write!(f, [text(","), soft_line_break_or_space(), entry.format()])?;
+        }
+        Ok(())
+    }
+}
+
+fn is_parenthesized(
+    range: TextRange,
+    elts: &[Expr],
+    f: &mut Formatter<PyFormatContext<'_>>,
+) -> bool {
+    let parentheses = "(";
+    let first_char = &f.context().contents()[TextRange::at(range.start(), parentheses.text_len())];
+    if first_char != parentheses {
+        return false;
+    }
+
+    // Consider `a = (1, 2), 3`: The first char of the current expr starts is a parentheses, but
```

Comment:
Is this true? I thought RustPython truncates the range of parenthesized expressions, so that they don't include the parentheses.

Reply:
Tuples are different from regular statements, their parentheses are included
```python
a = ((1), 2, 3)
```
The outer range (a tuple) is 4..15, the inner range (no tuple) is 6..7 .